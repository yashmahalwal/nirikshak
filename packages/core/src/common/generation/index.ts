export { generateFaker } from "./fakerGen";
export { generateCustom } from "./customGen";
