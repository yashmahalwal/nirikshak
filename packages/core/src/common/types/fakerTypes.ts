import _ from "lodash";
import faker from "faker";
import { FakerString } from "./fakerStrings";

// Type guard: Faker string
export function isFakerString(input: any): input is FakerString {
    if (typeof input !== "string") return false;

    const i = input.indexOf("faker:");
    if (i != 0) return false;

    return _.get(faker, input.slice(6)) instanceof Function;
}

// Faker object type
export type FakerObject = {
    function: FakerString;
    args?: any[];
};

// type guard: Faker object
export function isFakerObject(input: any): input is FakerObject {
    if (!input || typeof input !== "object") return false;

    if ("function" in input && isFakerString(input["function"])) {
        if ("args" in input) {
            return Array.isArray(input["args"]);
        }
        return true;
    }

    return false;
}

// The faker literal can be either of both
export type FakerType = FakerString | FakerObject;

// Type guard: faker type
export function isFakerType(input: any): input is FakerType {
    return isFakerString(input) || isFakerObject(input);
}

// A standard format to rep faker types
// Same as faker object, but optional fields are made compulsory
export type NormalizedFaker = {
    [K in keyof FakerObject]-?: FakerObject[K];
};

// Function to normalize faker literal
export function normalizeFaker(input: FakerType): NormalizedFaker {
    if (isFakerObject(input)) return { args: [], ...input };
    return {
        function: input,
        args: [],
    };
}
