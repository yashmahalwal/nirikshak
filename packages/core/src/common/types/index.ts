export { SchemaHelpers } from "./helpers";
export { isCustomFunction, CustomFunction } from "./custom";
export { isFakerType } from "./fakerTypes";
