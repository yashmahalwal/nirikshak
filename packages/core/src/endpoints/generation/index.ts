export { generateURL } from "./urlStringGen";
export { getFromResource } from "./resourceStringGen";
export * from "./helpers";
export * from "./input";
