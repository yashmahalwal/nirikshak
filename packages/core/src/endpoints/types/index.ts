export {
    MethodType,
    HeaderMap,
    Cases,
    BodyType,
    BodyInstance,
} from "./helpers";
export { isURLString } from "./urlString";
export { isResourceString } from "./resourceString";
export { isDescription, Description } from "./description";
export { HeaderAndQuery, Inputs } from "./input";
export { HeaderAndStatus, Outputs } from "./output";
