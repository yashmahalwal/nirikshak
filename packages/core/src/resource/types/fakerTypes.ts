import _ from "lodash";
import faker from "faker";

type FakerString = string;

function isValidFakerString(input: any): input is FakerString {
  if (typeof input !== "string") return false;

  const i = input.indexOf("faker:");
  if (i != 0) return false;

  return _.get(faker, input.slice(i + 6)) instanceof Function;
}

type FakerObject = {
  function: FakerString;
  args?: any[];
};

function isValidFakerObject(input: any): input is FakerObject {
  if (!input || typeof input !== "object") return false;

  if ("function" in input && isValidFakerString(input["function"])) {
    if ("args" in input) {
      return Array.isArray(input["args"]);
    }
    return true;
  }

  return false;
}

export type FakerType = FakerString | FakerObject;

function isValidFaker(input: any): input is FakerType {
  return isValidFakerString(input) || isValidFakerObject(input);
}

function normalizeFaker(input: FakerType): FakerObject {
  if (isValidFakerObject(input)) return input;
  return {
    function: input,
    args: [],
  };
}
