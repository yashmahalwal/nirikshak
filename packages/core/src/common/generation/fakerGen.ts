import { FakerType, normalizeFaker } from "../types/fakerTypes";
import faker from "faker";
import _ from "lodash";
import { Literal } from "../types/literals";

// Function to generate faker literals
// Should not throw any error for valid faker types
export function generateFaker(input: FakerType): Literal {
    const object = normalizeFaker(input);

    const functionName = object.function.slice(6);
    return _.get(faker, functionName)(...object.args);
}
