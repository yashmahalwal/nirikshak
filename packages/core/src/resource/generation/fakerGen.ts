import { FakerType, normalizeFaker } from "../../common/types/fakerTypes";
import faker from "faker";
import _ from "lodash";
import { Primitives } from "../types/helper";

// Function to generate faker literals
// Should not throw any error for valid faker types
export function generateFaker(input: FakerType): Primitives {
    const object = normalizeFaker(input);

    const functionName = object.function.slice(6);
    return _.get(faker, functionName)(...object.args);
}
