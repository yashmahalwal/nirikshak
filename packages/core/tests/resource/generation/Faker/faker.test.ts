/* 
    Tests to validate faker data generation validation
*/
import { FakerType } from "../../../../src/common/types/fakerTypes";
import faker from "faker";
import { generateFaker } from "../../../../src/resource/generation/fakerGen";
const ValidFakers: FakerType[] = [
    { function: "faker:address.zipCode", args: ['"#####-###"'] },
    "faker:address.citySuffix",
    "faker:address.streetName",
    "faker:address.streetAddress",
    "faker:address.streetPrefix",
    "faker:address.county",
    "faker:address.countryCode",
    { function: "faker:address.state", args: [true] },
    "faker:commerce.productAdjective",
    "faker:commerce.product",
    { function: "faker:company.bs", args: [] },
    {
        function: "faker:company.catchPhraseAdjective",
        args: [null, null, null],
    },
    "faker:company.catchPhraseDescriptor",
    "faker:company.bsNoun",
    "faker:database.collation",
    { function: "faker:finance.account", args: [8] },
    "faker:hacker.verb",
    "faker:image.image",
    "faker:image.avatar",
    { function: "faker:image.cats", args: [200, 200] },
    { function: "faker:internet.userName", args: ["FirstName", "LastName"] },
    "faker:internet.url",
    { function: "faker:lorem.word", args: [3] },
    "faker:lorem.sentence",
    { function: "faker:lorem.slug", args: [4] },
    "faker:lorem.paragraph",
    "faker:name.firstName",
    "faker:name.lastName",
    {
        function: "faker:random.number",
        args: [{ min: 0, max: 1, precision: 0.1 }],
    },
    "faker:random.words",
    "faker:system.semver",
];

describe("Faker literal generation", () => {
    // Setting randomness seed
    // Does not work with random.uuid and Date.*
    // So avoid them during testing
    beforeAll(() => faker.seed(123));

    test.each(ValidFakers)(`Valid faker: %#`, (entry) =>
        expect(generateFaker(entry)).toMatchSnapshot()
    );
});
