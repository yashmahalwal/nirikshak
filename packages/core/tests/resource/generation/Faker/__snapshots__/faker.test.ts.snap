// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Faker literal generation Valid faker: 0 1`] = `"\\"67242-657\\""`;

exports[`Faker literal generation Valid faker: 1 1`] = `"chester"`;

exports[`Faker literal generation Valid faker: 2 1`] = `"Satterfield Meadows"`;

exports[`Faker literal generation Valid faker: 3 1`] = `"465 Hudson Manors"`;

exports[`Faker literal generation Valid faker: 4 1`] = `"b"`;

exports[`Faker literal generation Valid faker: 5 1`] = `"Berkshire"`;

exports[`Faker literal generation Valid faker: 6 1`] = `"MW"`;

exports[`Faker literal generation Valid faker: 7 1`] = `"Oregon"`;

exports[`Faker literal generation Valid faker: 8 1`] = `"Incredible"`;

exports[`Faker literal generation Valid faker: 9 1`] = `"Table"`;

exports[`Faker literal generation Valid faker: 10 1`] = `"24/7 streamline infomediaries"`;

exports[`Faker literal generation Valid faker: 11 1`] = `"Integrated"`;

exports[`Faker literal generation Valid faker: 12 1`] = `"maximized"`;

exports[`Faker literal generation Valid faker: 13 1`] = `"mindshare"`;

exports[`Faker literal generation Valid faker: 14 1`] = `"ascii_general_ci"`;

exports[`Faker literal generation Valid faker: 15 1`] = `"14105754"`;

exports[`Faker literal generation Valid faker: 16 1`] = `"calculate"`;

exports[`Faker literal generation Valid faker: 17 1`] = `"http://lorempixel.com/640/480/cats"`;

exports[`Faker literal generation Valid faker: 18 1`] = `"https://s3.amazonaws.com/uifaces/faces/twitter/sweetdelisa/128.jpg"`;

exports[`Faker literal generation Valid faker: 19 1`] = `"http://lorempixel.com/200/200/cats"`;

exports[`Faker literal generation Valid faker: 20 1`] = `"FirstName_LastName"`;

exports[`Faker literal generation Valid faker: 21 1`] = `"https://leonor.info"`;

exports[`Faker literal generation Valid faker: 22 1`] = `"quisquam"`;

exports[`Faker literal generation Valid faker: 23 1`] = `"Quia expedita commodi autem tempora ut."`;

exports[`Faker literal generation Valid faker: 24 1`] = `"exercitationem-doloribus-harum-voluptas"`;

exports[`Faker literal generation Valid faker: 25 1`] = `"Et accusantium et quia quos sequi molestiae beatae enim. Molestiae placeat saepe eligendi rerum excepturi et laudantium fuga similique. Corporis voluptatem dolores."`;

exports[`Faker literal generation Valid faker: 26 1`] = `"Harvey"`;

exports[`Faker literal generation Valid faker: 27 1`] = `"Parker"`;

exports[`Faker literal generation Valid faker: 28 1`] = `0.9`;

exports[`Faker literal generation Valid faker: 29 1`] = `"Configurable Berkshire Manager"`;

exports[`Faker literal generation Valid faker: 30 1`] = `"5.4.6"`;
